### PR TITLE
Increase the maximum of ancient slots allowed to be not combined

### PR DESCRIPTION
**Pull request by dmakarov**
#### Problem

Large files containing ancient slots are expensive to update when the contained account data is being overwritten in more recent slots.

#### Summary of Changes

We increase the value of the tuning parameter `max_ancient_slots` to allow more ancient slots to remain in their individual storages instead of being combined into larger files. The smaller files are faster to update at the expense of the validator using more file handles.

### CODE REVIEW

**Comment by dmakarov** [comment]
I'm working on adding an option to set this parameter on the validator's command line.

**Comment by jeffwashington** [comment]
I suggest coupling this change with:
`ANCIENT_APPEND_VEC_DEFAULT_OFFSET = 100_000`

This keeps the overall 'ancient' range at 100k max, and the overall storage range at 432k, so we are at status quo relative today regarding # slots and # file handles.

**Review by HaoranYi — DISMISSED** [review]
max file handle = 430K + 100K + 100K = 630K. 
not too bad. 
lgtm!

**Comment by jeffwashington** [comment]
> max file handle = 430K + 100K + 100K = 630K. not too bad. lgtm!

I think 100k means move it into the 432k. I think -10k moves it past the end (so max of 432k - - 10k = 442k.

I'm proposing 432k - 100k + 100k = 432k total.

**Comment by HaoranYi** [comment]
Ah. I see. 

So we will update this in the next PR?

https://github.com/anza-xyz/agave/blob/369d2cb2ab958d852c1d9699b6a495a415675f25/accounts-db/src/accounts_db.rs#L598

**Comment by jeffwashington** [comment]
> Ah. I see.
> 
> So we will update this in the next PR?
> 
> https://github.com/anza-xyz/agave/blob/369d2cb2ab958d852c1d9699b6a495a415675f25/accounts-db/src/accounts_db.rs#L598

I wanted that done IN this pr:

https://github.com/anza-xyz/agave/pull/3205#issuecomment-2420230122

I guess I could've been more clear.

**Comment by dmakarov** [comment]
> Ah. I see.
> 
> So we will update this in the next PR?
> 
> https://github.com/anza-xyz/agave/blob/369d2cb2ab958d852c1d9699b6a495a415675f25/accounts-db/src/accounts_db.rs#L598

in this one.

File: accounts-db/src/accounts_db.rs
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: i think this is + 100_000

Comment (dmakarov) [reply accounts-db/src/accounts_db.rs]: shouldn't it be an offset to smaller slot numbers?  if the offset is positive, the logic of several tests has to be adjusted at least, maybe some logic in the algorithms as well. let me figure this out.

**Comment by jeffwashington** [comment]
we definitely need to start a skipping rewrites version of this and it needs to run more than an epoch. Skipping rewrites is still opt-in and the current master behavior is unacceptable, so even if this isn't completely right, the risk is low at the moment.

**Comment by dmakarov** [comment]
maybe I should merge #3206, rebase this on new master, and then start a skipping rewrites validator? so that all tweaks are included...

**Comment by jeffwashington** [comment]
> maybe I should merge #3206, rebase this on new master, and then start a skipping rewrites validator? so that all tweaks are included...

@dmakarov yes, I just tracked these down. 3206 is approved. I say merge it.

File: accounts-db/src/accounts_db.rs
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:617]: Should `MAX_ANCIENT_SLOTS_DEFAULT` be used here?

```suggestion
const ANCIENT_APPEND_VEC_DEFAULT_OFFSET: Option<i64> = Some(MAX_ANCIENT_SLOTS_DEFAULT);
```

Do we need two constants?

Comment (dmakarov) [reply accounts-db/src/accounts_db.rs:617]: i think they may have different values, in theory...

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:617]: Ok, I think I'm fine with two constants. I still think that by default we probably want the offset to be the same as the max number of ancient slots.

File: accounts-db/src/accounts_db.rs
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: nit: I find it confusing to have a variable named `root0` but not have a value of 0. Often we have `slotX` or `rootX` where the `X` is the slot number. Here we're breaking that pattern, and thus adds unexpectedness (which then I try and figure out "why"). I think we could rename the variable and bypass this issue entirely.

File: accounts-db/src/accounts_db.rs
Comment (dmakarov) [reply accounts-db/src/accounts_db.rs:617]: i believe they’re not related, that is changing one value shouldn’t affect the other. otherwise why should there be two constants, instead of one? Perhaps, follow up changes should reconsider the tuning parameters altogether.

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:617]: Yeah, that's fair. I guess I was under the impression that by setting the offset to a positive value, and setting the max number of ancient slots be the same value, that we were implicitly trying to communicate that we wanted a maximum number of storages to be about the number of slots per epoch.

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:617]: Can we include the discussion here in the code comments, i.e. document the relationship between these two constants?

Comment (dmakarov) [reply accounts-db/src/accounts_db.rs:617]: by the way max_ancient_slots seems to be a misnomer, because the value designates the number of storage files, not slots (there are in general accounts data from more slots packed into this number of storage files, if I understand correctly).

Comment (dmakarov) [reply accounts-db/src/accounts_db.rs:617]: I added a note to the comment explaining my understanding of the relationship between these two constants. Please, correct me if my understanding is wrong.

**Comment by dmakarov** [comment]
Is there something we're waiting for this PR? I think I addressed all the comments.

File: accounts-db/src/accounts_db.rs
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:617]: A positive value here will enable ancient storages/packing by default on master, without any additional hidden/testing CLI args. We probably need to keep this default as a negative value for now, until skipping rewrites is enabled.

Comment (dmakarov) [reply accounts-db/src/accounts_db.rs:617]: This concern is handled in #3266

**Review by HaoranYi — APPROVED** [review]
lgtm

**Review by brooksprumo — COMMENTED** [review]
Looks good to me. Happy to approve once #3266 is merged.

**Comment by dmakarov** [comment]
> Looks good to me. Happy to approve once #3266 is merged.

it's merged.

**Comment by jeffwashington** [comment]
> Is there something we're waiting for this PR? I think I addressed all the comments.

My testing found that we can get stuck when I enabled this change. We are not reducing alive_bytes when we encounter multi ref storages we can't pack yet.
We also aren't shrinking every ancient storage with dead bytes - just the ones < 90% alive.
I think both of those need to go in or we may find ourselves in a looming oom situation like I did where we fail to make progress.

**Comment by dmakarov** [comment]
> > Is there something we're waiting for this PR? I think I addressed all the comments.
> 
> My testing found that we can get stuck when I enabled this change. We are not reducing alive_bytes when we encounter multi ref storages we can't pack yet. We also aren't shrinking every ancient storage with dead bytes - just the ones < 90% alive. I think both of those need to go in or we may find ourselves in a looming oom situation like I did where we fail to make progress.

so this depends on #3251 and #3252 then.

**Review by jeffwashington — APPROVED** [review]
lgtm. dependencies are merged

**Pull request merged**
